### PR TITLE
Add support for base dilation and window dilation to reduce window op…

### LINKED ISSUE #3793: Make `lax.reduce_window` support dilation
It would be very convenient to have `lax.reduce_window` (https://github.com/google/jax/blob/fa2a0275c83c03cfad9d36f5da06b0bf47eedfb9/jax/lax/lax.py#L1087) support dilation specification, like https://www.tensorflow.org/xla/operation_semantics#reducewindow

Thanks!

### PR DESCRIPTION
**Pull request by hawkinsp**
…erators.

Refactor reduce-window tests to use many generated cases instead of a single large test case with a loop.

Treat `inf - inf` as a zero difference in gradient tests.

Fixes #3793

### CODE REVIEW

File: jax/lax/lax.py [hunk 1]
```diff
@@ -4612,12 +4682,24 @@ def _common_reduce_window_shape_rule(operand, window_dimensions,
...
+    raise TypeError(msg.format(window_dilation, window_dimensions))
 
   return reduce_window_shape_tuple(operand.shape, window_dimensions,
-                                   window_strides, padding)
+                                   window_strides, padding, base_dilation,
+                                   window_dilation)
 
 def reduce_window_shape_tuple(operand_shape, window_dimensions, window_strides,
-                              padding):
+                              padding, base_dilation, window_dilation):
```
Comment (jekbradbury) [inline jax/lax/lax.py:4700]: If we had defaults for `base_dilation` and `window_dilation` here, we could avoid having to make downstream changes in the two consumers of `reduce_window_shape_tuple` (neural tangents and oryx). But default args are also bad, so maybe it's not worth it?

**Review by jekbradbury — APPROVED** [review]

File: jax/lax/lax.py (hunk 1 above)
Comment (hawkinsp) [reply jax/lax/lax.py:4700]: True.

Actually what we should do here is convince Neural Tangents and Oryx to use `jax.eval_shape` instead, rather than this undocumented API...?

I'll add the defaults for now.

Comment (hawkinsp) [reply jax/lax/lax.py]: Filed https://github.com/google/neural-tangents/issues/51 and emailed for Oryx.

Comment (jekbradbury) [reply jax/lax/lax.py]: Looks like it's also used in Flax; I filed https://github.com/google/flax/issues/370.

**Comment by llCurious** [comment]
hey @hawkinsp . i just wonder what is the application scenario of dilation in reduce window. Is the dilation here similar to convolution dilation?
